implement POST `/eth/v1/beacon/pool/sync_committees`

rolfyone opened the pull request:
- in order to support the altair milestone, validators need to be able to submit sync committee signatures.

 - currently marked experimental as the standard api spec is not finalized.

fixes #3906

Signed-off-by: Paul Harris <paul.harris@consensys.net>

## Documentation

- [X] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [X] I thought about adding a changelog entry, and added one if I deemed necessary.

rolfyone committed:
implement POST `/eth/v1/beacon/pool/sync_committees`

 - in order to support the altair milestone, validators need to be able to submit sync committee signatures.

 - currently marked experimental as the standard api spec is not finalized.

fixes #3906

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 3906

rolfyone committed:
spotless

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
add integration test to demonstrate errors coming out

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
fix build

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 3906

rolfyone committed:
implement the addition of sync committee signatures to the pool, and filtering rejections to return back to caller

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton commented on CHANGELOG.md:
```diff
@@ -15,6 +15,7 @@ For information on changes in released versions of Teku, see the [releases page]
 ### Additions and Improvements
 - Newly created databases will now use LevelDB for storage instead of RocksDB. This uses less memory and has proven to be more stable. Existing databases are unaffected and will continue to use RocksDB.
 - Support for automatic fail-over of eth1-endpoints.  Multiple endpoints can be specified with the new `--eth1-endpoints` CLI option. Thanks to Enrico Del Fante.
+- implement POST `/eth/v1/beacon/pool/sync_committees` to allow validators to submit sync committee signatures to the beacon node.
```
Given the functionality that actually makes this work doesn't exist yet it's probably too early to add to the changelog.

rolfyone replied:
i did add the functionality, just not sure if we call it working if we're not in altair?

ajsutton replied:
Oh right - I was thinking of the bit of code that still has a `// FIXME`. I think it's fine to add.  I'll probably wind up summarising them as just a single "added endpoints in preparation for Altair" kind of thing in the final release notes but no harm in having the full details here so we can decide later.

ajsutton commented on data/serializer/src/main/java/tech/pegasys/teku/api/schema/altair/SyncCommitteeSignature.java:
```diff
@@ -0,0 +1,86 @@
+        SszUInt64.of(slot),
+        SszBytes32.of(beaconBlockRoot),
+        SszUInt64.of(validatorIndex),
+        new SszSignature(signature.asInternalBLSSignature()));
```
hmm, this is somewhat unfortunate.  Ideally we'd go through `spec.getSchemaDefinitions` to create this.  Otherwise we're just hard coding specific schemas in more places.  I'm not 100% sure what the right trade off here is - specifically we need to handle the case where the slot is actually in phase0 and we can't create an internal version properly.

ajsutton commented on validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandler.java:
```diff
@@ -475,6 +484,47 @@ public void sendAggregateAndProof(final SignedAggregateAndProof aggregateAndProo
   }
 
+  @Override
+  public SafeFuture<Optional<SubmitCommitteeSignaturesResult>> sendSyncCommitteeSignatures(
```
We probably should add some tests for these new methods.

ajsutton commented on validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/ValidatorApiHandler.java:
```diff
@@ -475,6 +484,47 @@ public void sendAggregateAndProof(final SignedAggregateAndProof aggregateAndProo
+            .flatMap(
+                validationResult ->
+                    this.fromInternalValidationResult(
+                        validationResult, internalValidationResults.indexOf(validationResult))
```
We probably don't want to use `.indexOf` to get the index here - we'll wind up repeatedly searching the list so if a significant number of invalid signatures are submitted this will become quite slow as it's O(n^2).  Better to just use an old-skool `for (int index = 0; index < internalValidationResults.size(); index++)` kind of loop so we have the index available.

ajsutton commented on validator/remote/src/main/java/tech/pegasys/teku/validator/remote/RemoteValidatorApiHandler.java:
```diff
@@ -284,6 +288,34 @@ public void sendSignedAttestation(
+                .sendSyncCommitteeSignatures(
+                    syncCommitteeSignatures.stream()
+                        .map(
+                            i ->
```
nit: I'd probably call this `signature` instead of `I`.

rolfyone replied:
i got sick of typing nice names, but i'll fix.

ajsutton commented on validator/remote/src/main/java/tech/pegasys/teku/validator/remote/apiclient/OkHttpValidatorRestApiClient.java:
```diff
@@ -250,6 +252,13 @@ public void subscribeToPersistentSubnets(final Set<SubnetSubscription> subnetSub
+  @Override
+  public Optional<PostSyncCommitteeFailureResponse> sendSyncCommitteeSignatures(
+      final List<SyncCommitteeSignature> syncCommitteeSignatures) {
+    // FIXME
```
Probably remove the FIXME comment from here.  We have the ticket still open to follow up on this anyway.

rolfyone replied:
this should actually be fairly trivial, i'll make the post.

rolfyone committed:
review feedback

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
use altair schema definition to create Sync committee signature object per review comments and discusison.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton commented on data/provider/src/main/java/tech/pegasys/teku/api/ValidatorDataProvider.java:
```diff
@@ -146,6 +149,14 @@ public void submitAttestation(Attestation attestation) {
+  public SafeFuture<Optional<SubmitCommitteeSignaturesResult>> submitCommitteeSignatures(
+      final List<SyncCommitteeSignature> signatures) {
+    return validatorApiChannel.sendSyncCommitteeSignatures(
+        signatures.stream()
+            .flatMap(signature -> signature.asInternalCommitteeSignature(spec).stream())
+            .collect(Collectors.toList()));
```
I don't think this works - it will silently ignore any signatures that are from before the Altair fork and then the index reported in any errors from other signatures will wind up being wrong.

Since it's a corner case, I'd be tempted to just reject the entire submission if any signatures are from a phase0 slot.

rolfyone replied:
not sure what the desired behaviour should be if out of N submissions, 1 is prior to altair... like you said it's a corner case, but I do kind of agree swallowing them isn't ideal. I can try to make it return errors for those specific cases... My initial theory was I didn't really mind, but I can see where you're coming from.

rolfyone replied:
I've gone with rejecting the entire submission. it should be very rare anyway, but I did add a regression test showing the behaviour.

rolfyone committed:
cause immediate request failure if a phase 0 slot is received in the request.

This would occur before any processing effectively, so all parts of the request will be rejected, there won't be specific messages for each item.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge branch 'master' into 3906

rolfyone committed:
Merge branch 'master' into 3906

rolfyone committed:
Merge branch 'master' into 3906

ajsutton reviewed (approved):
LGTM.

rolfyone committed:
fix hidden merge conflict

Signed-off-by: Paul Harris <paul.harris@consensys.net>
